### PR TITLE
Have AsyncSpec use locks instead of sleeps

### LINKED ISSUE #195: polysemy-1.0.0.0 flaky test?
Just wanted to report that I saw the following test failure, but it did not fail again when I reran the test suite.

```
      test/AsyncSpec.hs:43:5: 
      1) Async.async should thread state and not lock
           ["1> hello","2> olleh","1> olleh"] does not contain ["1> hello","2> olleh","1> pong"]
```

This was on the stackage build server while building `nightly-2019-07-31`.

### PR DESCRIPTION
**Pull request by KingoftheHomeless**
This presumably fixes #195, but that issue warrants some further investigation

### CODE REVIEW

File: test/AsyncSpec.hs [hunk 1]
```diff
@@ -21,22 +22,24 @@ spec = describe "async" $ do
...
           v <- get @String
           message 1 v
           put $ reverse v
 
-          embed $ threadDelay 1e5
+          embed $ putMVar lock1 ()
+          embed $ takeMVar lock2
           get >>= message 1
 
-          embed $ threadDelay 1e5
```
Comment (KingoftheHomeless) [inline test/AsyncSpec.hs]: I don't get what this `threadDelay` is for, so I simply removed it.

Comment (isovector) [reply test/AsyncSpec.hs]: This was so that the final `get` in this block would read the `pong` from the other thread.

File: test/AsyncSpec.hs [hunk 2]
```diff
@@ -21,22 +22,24 @@ spec = describe "async" $ do
       let message :: Member Trace r => Int -> String -> Sem r ()
           message n msg = trace $ mconcat
             [ show n, "> ", msg ]
-
+      ~[lock1, lock2] <- embed $
```
Comment (isovector) [inline test/AsyncSpec.hs]: ooc, why the lazy match here?

Comment (KingoftheHomeless) [reply test/AsyncSpec.hs]: Irrefutable pattern match; doesn't require `Sem r` to be `MonadFail`.

File: test/AsyncSpec.hs (hunk 1 above)
Comment (KingoftheHomeless) [reply test/AsyncSpec.hs]: But this is already guaranteed (or should be, rather) from the previous `threadDelay`, isn't it?

File: test/AsyncSpec.hs (hunk 2 above)
Comment (isovector) [reply test/AsyncSpec.hs]: Right -- makes sense! Thanks for the explanation.

File: test/AsyncSpec.hs (hunk 1 above)
Comment (isovector) [reply test/AsyncSpec.hs]: The logic I intended (perhaps not what I implemented) was this interleaving of threads:

thread1
thread2
thread1

Comment (KingoftheHomeless) [reply test/AsyncSpec.hs]: Yeah, that is what happens:
1. thread2 sleeps. thread1 gets, messages, puts, and sleeps
2. thread2 wakes, gets, messages, puts, and terminates
3. thread1 wakes, gets, messages, puts, and sleeps.
4. thread1 wakes, gets, and terminates.
5. main thread completes its `await` as thread1 has terminated, puts, and terminates.

The `threadDelay` I removed was the one in time slot 3, as it's seemingly redundant.

**Comment by isovector** [comment]
Thanks!

**Pull request merged**